### PR TITLE
Internal stock requisition

### PR DESCRIPTION
**Pull request by mbayopanda**
This PR implement the internal stock requisition for : 
- Service asking depot
- Depot asking depot

WFvN1cNKBi

In this PR we can perform stock exit from a requisition for services and depots : 

VFuyo5y5T7

### CODE REVIEW

File: client/src/js/components/bhServiceOrDepot/bhServiceOrDepot.js
Comment (jniles) [inline client/src/js/components/bhServiceOrDepot/bhServiceOrDepot.js]: Can we fire all these at once?

```js
Promise.all([
  Stock.stockRequestorType.read(),
  Depots.read(),
  Services.read(),
]).then(([requestors, depots, services ]) => {
  Object.assign($ctrl, { requestors, depots, services});
  // etc ..
});

File: server/controllers/stock/reports/stock_exit_depot.receipt.handlebars
Comment (jniles) [inline server/controllers/stock/reports/stock_exit_depot.receipt.handlebars:9]: :+1:

File: server/models/migrations/next/migrate.sql
Comment (jniles) [inline server/models/migrations/next/migrate.sql]: Since you already did a bit of work here, go ahead and remove this stored procedure.  We don't need them in the migration file since we rebuild them from the repository source each time.

File: client/src/i18n/en/stock.json
Comment (jniles) [inline client/src/i18n/en/stock.json]: The `DOES_REQUISITION_EXIST_HELP` should read "Click on this option if you want to exit stock based on a requisition slip".

File: client/src/modules/stock/exit/exit.js
Comment (jniles) [inline client/src/modules/stock/exit/exit.js:258]: :dancer:

File: client/src/modules/stock/exit/exit.js
Comment (jniles) [inline client/src/modules/stock/exit/exit.js]: This is is a promise, correct? Should we be waiting for it to complete before showing the modal?  For example:
```js
let completedUuid;
return Stock.movements.create(movement)
  .then(document => {
    completedUuid = document.uuid;
    if (vm.requisition) {
      return Stock.stockRequisition.update(vm.requisition.uuid, { status_id : COMPLETED_STATUS });
    }
  })
  .then(() => {
    ReceiptModal.stockExitServiceReceipt(completedUuid, bhConstants.flux.TO_SERVICE
    // etc....
  });

File: client/src/js/components/bhHasRequisitionVoucher/bhHasRequisitionVoucher.html
Comment (jniles) [inline client/src/js/components/bhHasRequisitionVoucher/bhHasRequisitionVoucher.html]: Let's call this "requisitionReference" or something similar to help us track what it is.

File: client/src/js/components/bhServiceOrDepot/bhServiceOrDepot.html
Comment (jniles) [inline client/src/js/components/bhServiceOrDepot/bhServiceOrDepot.html]: This is cool 👍 .

Did you know that you can [translate attributes like this in a directive](https://angular-translate.github.io/docs/#/api/pascalprecht.translate.directive:translate-attr)?
```html
<label
  translate-attr="{ title : requestor.title_key  }">
```

File: client/src/js/components/bhServiceOrDepot/bhServiceOrDepot.js
Comment (jniles) [inline client/src/js/components/bhServiceOrDepot/bhServiceOrDepot.js]: It looks like [Array.prototype.some()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) is a better fit here, since you are just trying to figure out if the element matches without using it later.
```js
// returns true if it found a match.
const foundInService = $ctrl.services.some(elt => elt.uuid === identifier);
```

Comment (jniles) [reply client/src/js/components/bhServiceOrDepot/bhServiceOrDepot.js]: Actually, it doesn't look like you use depots or services at all, besides for this check.  If that is true, instead of keeping around an array of depots and services, can you just keep their uuids?  Like this:
```js
const $ctrl.depotIds = depots.map(depot => depot.uuid);

// ... later ...
const foundInDepot = $ctrl.depotIds.includes(identifier);
```
This would probably be a lot faster and make the code easier to read.

File: client/src/modules/stock/requisition/modals/action.modal.js
Comment (jniles) [inline client/src/modules/stock/requisition/modals/action.modal.js]: Do you really use depots in this controller/template?  It doesn't look like it to me.  If so, please take a second look at this controller and make sure you've removed all the items you don't need.

File: client/src/modules/stock/requisition/modals/action.modal.js
Comment (jniles) [inline client/src/modules/stock/requisition/modals/action.modal.js:94]: Can you put a loading indicator on the grid while you make this request?  That would let the user know something is happening.

Comment (jniles) [reply client/src/modules/stock/requisition/modals/action.modal.js:94]: You could even make the `autoSuggestInventories` button behave like the `bh-loading-button` and be disabled and show a loading indicator after you have clicked it as well for style points.

File: client/src/modules/stock/requisition/modals/action.modal.js
Comment (jniles) [inline client/src/modules/stock/requisition/modals/action.modal.js:106]: Nice.  I like it 👍

File: server/controllers/stock/requisition/requisition.js
Comment (jniles) [inline server/controllers/stock/requisition/requisition.js]: This should be [204](https://github.com/IMA-WorldHealth/bhima/wiki/Design:-Server-CRUD-APIs#api-design) - no content.

File: server/models/migrations/next/migrate.sql
Comment (jniles) [inline server/models/migrations/next/migrate.sql]: No need to put the triggers here.  We rebuild them from the base repo each time we upgrade.

**Review by jniles — COMMENTED** [review]
@mbayopanda, the functionality looks really good.  I've made a few comments to tighten up the code, but this is over 95% of the way there.

**Comment by jniles** [comment]
@mbayopanda, let me know when you need another review of this.  I remember it be close to complete and we can probably merge it pretty quickly.

**Comment by mbayopanda** [comment]
@jniles could you review this PR again please

**Review by jniles — APPROVED** [review]
LGTM.  I think this has a merge issue, but once that is fixed, it can be merged anytime.

**Comment by jniles** [comment]
Merge issue is fixed.

bors r+

**Pull request merged**
